File: backend_db/SDKs/swift/Sources/SelfDB/Storage.swift
import Foundation

/// Files operations
public struct Files: Sendable {
    private let client: HTTPClient
    
    init(client: HTTPClient) {
        self.client = client
    }
    
    /// Get storage statistics
    /// GET /storage/files/stats
    public func stats() async throws -> StorageStatsResponse {
        return try await client.get(path: "storage/files/stats")
    }
    
    /// Get total file count
    /// GET /storage/files/total-count
    public func totalCount(search: String? = nil) async throws -> Int {
        var params: [String: String] = [:]
        if let search = search { params["search"] = search }
        
        let response: CountResponse = try await client.get(path: "storage/files/total-count", queryParams: params.isEmpty ? nil : params)
        return response.count
    }
    
    /// Get file count for a bucket
    /// GET /storage/files/count
    public func count(bucketId: String) async throws -> Int {
        let params = ["bucket_id": bucketId]
        let response: CountResponse = try await client.get(path: "storage/files/count", queryParams: params)
        return response.count
    }
    
    /// Upload a file to a bucket
    /// POST /storage/files/upload
    public func upload(
        _ bucketId: String,
        filename: String,
        data: Data,
        path: String? = nil
    ) async throws -> FileUploadResponse {
        var params: [String: String] = [
            "bucket_id": bucketId,
            "filename": filename
        ]
        if let path = path { params["path"] = path }
        
        return try await client.postRaw(path: "storage/files/upload", body: data, queryParams: params, contentType: "application/octet-stream")
    }
    
    /// List files with optional filtering
    /// GET /storage/files/
    public func list(
        bucketId: String? = nil,
        page: Int = 1,
        pageSize: Int = 100,
        search: String? = nil
    ) async throws -> FileDataResponse {
        var params: [String: String] = [
            "page": String(page),
            "page_size": String(pageSize)
        ]
        if let bucketId = bucketId { params["bucket_id"] = bucketId }
        if let search = search { params["search"] = search }
        
        return try await client.get(path: "storage/files/", queryParams: params)
    }
    
    /// Get a file by ID
    /// GET /storage/files/{file_id}
    public func get(_ fileId: String) async throws -> FileResponse {
        return try await client.get(path: "storage/files/\(fileId)")
    }
    
    /// Delete a file
    /// DELETE /storage/files/{file_id}
    public func delete(_ fileId: String) async throws {
        try await client.deleteEmpty(path: "storage/files/\(fileId)")
    }
    
    /// Update file metadata
    /// PATCH /storage/files/{file_id}
    public func updateMetadata(_ fileId: String, metadata: [String: AnyCodable]) async throws -> FileResponse {
        struct MetadataUpdate: Encodable {
            let metadata: [String: AnyCodable]
        }
        return try await client.patch(path: "storage/files/\(fileId)", body: MetadataUpdate(metadata: metadata))
    }
    
    /// Download a file
    /// GET /storage/files/download/{bucket_name}/{path}
    public func download(bucketName: String, path: String) async throws -> Data {
        let encodedPath = path.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? path
        return try await client.getRaw(path: "storage/files/download/\(bucketName)/\(encodedPath)")
    }
}

/// Buckets operations
public struct Buckets: Sendable {
    private let client: HTTPClient
    
    init(client: HTTPClient) {
        self.client = client
    }
    
    /// Get bucket count
    /// GET /storage/buckets/count
    public func count(search: String? = nil) async throws -> Int {
        var params: [String: String] = [:]
        if let search = search { params["search"] = search }
        
        let response: CountResponse = try await client.get(path: "storage/buckets/count", queryParams: params.isEmpty ? nil : params)
        return response.count
    }
    
    /// Create a new bucket
    /// POST /storage/buckets/
    public func create(payload: BucketCreate) async throws -> BucketResponse {
        return try await client.post(path: "storage/buckets/", body: payload)
    }
    
    /// List buckets with optional filtering
    /// GET /storage/buckets/
    public func list(
        skip: Int = 0,
        limit: Int = 100,
        search: String? = nil,
        sortBy: BucketSortBy? = nil,
        sortOrder: SortOrder? = nil
    ) async throws -> [BucketResponse] {
        var params: [String: String] = [
            "skip": String(skip),
            "limit": String(limit)
        ]
        if let search = search { params["search"] = search }
        if let sortBy = sortBy { params["sort_by"] = sortBy.rawValue }
        if let sortOrder = sortOrder { params["sort_order"] = sortOrder.rawValue }
        
        return try await client.get(path: "storage/buckets/", queryParams: params)
    }
    
    /// Get a bucket by ID
    /// GET /storage/buckets/{bucket_id}
    public func get(_ bucketId: String) async throws -> BucketResponse {
        return try await client.get(path: "storage/buckets/\(bucketId)")
    }
    
    /// Update a bucket
    /// PATCH /storage/buckets/{bucket_id}
    public func update(_ bucketId: String, payload: BucketUpdate) async throws -> BucketResponse {
        return try await client.patch(path: "storage/buckets/\(bucketId)", body: payload)
    }
    
    /// Delete a bucket
    /// DELETE /storage/buckets/{bucket_id}
    public func delete(_ bucketId: String) async throws {
        try await client.deleteEmpty(path: "storage/buckets/\(bucketId)")
    }
}

/// Storage module for bucket and file management
public struct Storage: Sendable {
    /// Buckets operations
    public let buckets: Buckets
    
    /// Files operations
    public let files: Files
    
    init(client: HTTPClient) {
        self.buckets = Buckets(client: client)
        self.files = Files(client: client)
    }
}

File: backend_db/SDKs/typescript/src/storage.ts
/**
 * SelfDB SDK Storage Module
 * 
 * Bucket and file management for storage operations.
 */

import { HttpClient } from './client';
import {
    BucketCreate,
    BucketUpdate,
    BucketResponse,
    FileUploadResponse,
    FileResponse,
    FileDataResponse,
    StorageStatsResponse,
    CountResponse,
    PaginationOptions,
} from './models';

/**
 * Buckets resource for bucket CRUD operations
 */
export class BucketsResource {
    constructor(private client: HttpClient) {}

    /**
     * Get bucket count
     * GET /storage/buckets/count
     */
    async count(options: { search?: string } = {}): Promise<CountResponse> {
        const query = options.search ? { search: options.search } : undefined;
        return this.client.get<CountResponse>('/storage/buckets/count', query);
    }

    /**
     * Create a bucket
     * POST /storage/buckets/
     */
    async create(bucket: BucketCreate): Promise<BucketResponse> {
        return this.client.post<BucketResponse>('/storage/buckets/', bucket);
    }

    /**
     * List buckets with optional pagination and search
     * GET /storage/buckets/
     */
    async list(options: PaginationOptions = {}): Promise<BucketResponse[]> {
        const query: Record<string, string | number | boolean | undefined> = {
            skip: options.skip,
            limit: options.limit,
            search: options.search,
            sort_by: options.sortBy,
            sort_order: options.sortOrder,
        };
        return this.client.get<BucketResponse[]>('/storage/buckets/', query);
    }

    /**
     * Get a bucket by ID
     * GET /storage/buckets/{bucket_id}
     */
    async get(bucketId: string): Promise<BucketResponse> {
        return this.client.get<BucketResponse>(`/storage/buckets/${bucketId}`);
    }

    /**
     * Update a bucket
     * PATCH /storage/buckets/{bucket_id}
     */
    async update(bucketId: string, updates: BucketUpdate): Promise<BucketResponse> {
        return this.client.patch<BucketResponse>(`/storage/buckets/${bucketId}`, updates);
    }

    /**
     * Delete a bucket
     * DELETE /storage/buckets/{bucket_id}
     */
    async delete(bucketId: string): Promise<void> {
        await this.client.delete<void>(`/storage/buckets/${bucketId}`);
    }
}

/**
 * Files resource for file operations
 */
export class FilesResource {
    constructor(private client: HttpClient) {}

    /**
     * Get storage statistics
     * GET /storage/files/stats
     */
    async stats(): Promise<StorageStatsResponse> {
        return this.client.get<StorageStatsResponse>('/storage/files/stats');
    }

    /**
     * Get total file count
     * GET /storage/files/total-count
     */
    async totalCount(options: { search?: string } = {}): Promise<CountResponse> {
        const query = options.search ? { search: options.search } : undefined;
        return this.client.get<CountResponse>('/storage/files/total-count', query);
    }

    /**
     * Get file count for a bucket
     * GET /storage/files/count
     */
    async count(options: { bucketId?: string; search?: string } = {}): Promise<CountResponse> {
        const query: Record<string, string | number | boolean | undefined> = {
            bucket_id: options.bucketId,
            search: options.search,
        };
        return this.client.get<CountResponse>('/storage/files/count', query);
    }

    /**
     * Upload a file to a bucket
     * POST /storage/files/upload
     * NOTE: Uses raw bytes body with query params (not multipart)
     */
    async upload(
        bucketId: string,
        options: {
            filename: string;
            data: ArrayBuffer | Uint8Array | Blob | string;
            path?: string;
            contentType?: string;
        }
    ): Promise<FileUploadResponse> {
        const query: Record<string, string | number | boolean | undefined> = {
            bucket_id: bucketId,
            filename: options.filename,
            path: options.path,
        };

        const headers: Record<string, string> = {};
        if (options.contentType) {
            headers['Content-Type'] = options.contentType;
        } else {
            headers['Content-Type'] = 'application/octet-stream';
        }

        // Convert data to appropriate format for fetch
        let body: Blob;
        if (typeof options.data === 'string') {
            body = new Blob([options.data]);
        } else if (options.data instanceof Blob) {
            body = options.data;
        } else if (options.data instanceof ArrayBuffer) {
            body = new Blob([options.data]);
        } else {
            // Uint8Array - create a new ArrayBuffer copy to avoid SharedArrayBuffer issues
            const copy = new Uint8Array(options.data).buffer;
            body = new Blob([copy]);
        }

        return this.client.postRaw<FileUploadResponse>('/storage/files/upload', body, query, headers);
    }

    /**
     * List files with optional pagination
     * GET /storage/files/
     */
    async list(options: {
        bucketId?: string;
        skip?: number;
        limit?: number;
        pageSize?: number;
        search?: string;
    } = {}): Promise<FileDataResponse> {
        const query: Record<string, string | number | boolean | undefined> = {
            bucket_id: options.bucketId,
            skip: options.skip,
            limit: options.limit,
            page_size: options.pageSize,
            search: options.search,
        };
        return this.client.get<FileDataResponse>('/storage/files/', query);
    }

    /**
     * Get a file by ID
     * GET /storage/files/{file_id}
     */
    async get(fileId: string): Promise<FileResponse> {
        return this.client.get<FileResponse>(`/storage/files/${fileId}`);
    }

    /**
     * Delete a file
     * DELETE /storage/files/{file_id}
     */
    async delete(fileId: string): Promise<void> {
        await this.client.delete<void>(`/storage/files/${fileId}`);
    }

    /**
     * Update file metadata
     * PATCH /storage/files/{file_id}
     */
    async updateMetadata(fileId: string, metadata: Record<string, unknown>): Promise<FileResponse> {
        return this.client.patch<FileResponse>(`/storage/files/${fileId}`, { metadata });
    }

    /**
     * Download a file by bucket name and path
     * GET /storage/files/download/{bucket_name}/{path}
     */
    async download(options: { bucketName: string; path: string }): Promise<ArrayBuffer> {
        const path = options.path.startsWith('/') ? options.path.slice(1) : options.path;
        return this.client.get<ArrayBuffer>(`/storage/files/download/${options.bucketName}/${path}`);
    }
}

/**
 * Storage module for bucket and file management
 */
export class Storage {
    public readonly buckets: BucketsResource;
    public readonly files: FilesResource;

    constructor(client: HttpClient) {
        this.buckets = new BucketsResource(client);
        this.files = new FilesResource(client);
    }
}
